File: scripts/combine-dense.jl
using DrWatson
@quickactivate "HubbardFTLM"

using HubbardFTLM
using DataFrames
using CSV
using ProgressMeter

using ArgParse
using Glob
using HDF5
using JSON3

function parse_commandline()
    s = ArgParse.ArgParseSettings()
    @add_arg_table! s begin
        "lattice"
            arg_type = String
            help = "shape of the lattice in the format lattice-(?,?)x(?,?)"
            required = true
    end
    parse_args(s)
end


function main()
    parsed_args = parse_commandline()
    # latticetype, shape = parse_lattice(parsed_args["lattice"])
    lattice_str = parsed_args["lattice"]
    dense_filepath = datadir(lattice_str, "eigen", "eigen-dense-results.hdf5")
    filenames = Glob.glob("temp-eigen-dense-*.jsonl", datadir(lattice_str, "eigen"))

    IdentifierType = NamedTuple{(:idx, :hopping, :interaction), Tuple{Int, Float64, Float64}}
    existing_results = Set(IdentifierType[])
    dense_hdf5 = h5open(dense_filepath, "cw")
    haskey(dense_hdf5, "eigen-dense") || create_group(dense_hdf5, "eigen-dense")
    h5g_eigen = dense_hdf5["eigen-dense"]

    for c in h5g_eigen
        idx = read(attributes(c)["idx"])
        t = read(attributes(c)["hopping"])
        U = read(attributes(c)["interaction"])
        push!(existing_results, (idx=idx, hopping=t, interaction=U))
    end

    @info "Read $dense_filepath with $(length(existing_results)) parameter sets."
    skipcount = 0
    for filename in filenames
        # df = CSV.read(filename, DataFrame)
        data = open(filename, "r") do io
            JSON3.read(io; jsonlines=true, allow_inf=false)
        end
        for item in data
            idx::Int = item["idx"]
            hopping::Float64 = item["hopping"]
            interaction::Float64 = item["interaction"]
            eigenvalues::Vector{Float64} = [item["eigenvalue"]...]
            githash::String = item["githash"]
            timestamp::String = item["timestamp"]

            identifier = (idx=idx, hopping=hopping, interaction=interaction)
            if identifier in existing_results
                skipcount += 1
                continue
            end
            group_name = savename((idx=idx, hopping=hopping, interaction=interaction))
            group = create_group(h5g_eigen, group_name)
            attributes(group)["idx"] = idx
            attributes(group)["hopping"] = hopping
            attributes(group)["interaction"] = interaction
            attributes(group)["githash"] = githash
            attributes(group)["timestamp"] = timestamp
            group["eigenvalue"] = eigenvalues
            flush(dense_hdf5)
            push!(existing_results, identifier)
        end
        @info "Appended $filename. (Skipped $skipcount entries)"
    end
    @info "Total $(length(h5g_eigen)) parameter sets"


    existing_results = Set(IdentifierType[])
    delete_list = String[]
    for c in h5g_eigen
        idx = read(attributes(c)["idx"])
        t = read(attributes(c)["hopping"])
        U = read(attributes(c)["interaction"])
        identifier = (idx=idx, hopping=t, interaction=U)
        if identifier in existing_results
            push!(delete_list, HDF5.name(c))
        end
        push!(existing_results, identifier)
    end

    @info "Delete list (not implemented yet)" delete_list

    close(dense_hdf5)
    for filename in filenames
        mv(filename, "$filename.processed")
    end
end

main()

File: scripts/combine-dense.py
import os, sys
import json
import numpy as np
import scipy as sp
import fastavro
import pandas as pd
import argparse
from pathlib import Path

project_dir = (Path(os.path.dirname(__file__)) / "..").resolve()

with (project_dir / "src" / "eigenvalues.avsc").open("r") as f:
    schema = json.load(f)
parsed_schema = fastavro.parse_schema(schema)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("lattice_str", type=str)
    args = parser.parse_args()
    filenames = (project_dir / "data" / args.lattice_str / "eigen").glob("*.jsonl")

    existing_parameters = set()
    result_filepath = (project_dir / "data" / args.lattice_str / "eigen" / "dense-results.avro")
    readflag = result_filepath.is_file()
    with result_filepath.open("ab+") as fi:
        fi.seek(0)
        if readflag:
            avro_reader = fastavro.reader(fi)
            for record in avro_reader:
                t = record["hopping"]
                U = record["interaction"]
                idx = record["idx"]
                existing_parameters.add((t, U, idx))
        else:
            fastavro.writer(fi, parsed_schema, [])
        
        for filename in filenames:
            df = pd.read_json(filename, lines=True)
            nr = len(df)
            df = df[df.apply(lambda row: (row.hopping, row.interaction, row.idx) not in existing_parameters, axis=1)]
            nr2 = len(df)
            print(f"{nr} -> keep {nr2} / throw {nr-nr2}")
            df["timestamp"] = (df.timestamp - pd.Timestamp(1970, 1,1)) // pd.Timedelta("1ms")
            df.apply(lambda row: existing_parameters.add((row.hopping, row.interaction, row.idx)), axis=1)
            fastavro.writer(fi, parsed_schema, df.to_dict('records'))
            os.rename(filename, f"{filename}.processed")
            

if __name__=='__main__':
    main()

